xchange-okcoin/src/main/java/info/bitrich/xchangestream/okcoin/OkCoinV3StreamingService.java

```diff
@@ -0,0 +1,155 @@
+
+  private Disposable pingPongSubscription;
+
+  public OkCoinV3StreamingService(String apiUrl) {
+    super(apiUrl);
+  }
+
+  @Override
+  public Completable connect() {
+    Completable conn = super.connect();
+    return conn.andThen(
+        (CompletableSource)
+            (completable) -> {
+              try {
+                if (pingPongSubscription != null && !pingPongSubscription.isDisposed()) {
+                  pingPongSubscription.dispose();
+                }
+                pingPongSubscription =
+                    pingPongSrc.subscribe(
+                        o -> {
+                          this.sendMessage("ping");
+                        });
+                completable.onComplete();
+              } catch (Exception e) {
+                completable.onError(e);
+              }
+            });
+  }
+
+  @Override
+  protected String getChannelNameFromMessage(JsonNode message) throws IOException {
+    JsonNode table = message.get("table");
+    if (table != null) {
+      JsonNode data = message.get("data");
+      JsonNode instrument_id = data.get(0).get("instrument_id");
+      String pair = instrument_id.toString();
+      String result = table + ":" + pair;
+      return result.replace("\"", "");
+    }
+    String event = message.get("event").asText();
+    return event;
+  }
+
+  @Override
+  public String getSubscribeMessage(String channelName, Object... args) throws IOException {
+
+    return String.format("{\"op\": \"subscribe\", \"args\": [\"%s\"]}", channelName);
+  }
+
+  @Override
+  public String getUnsubscribeMessage(String channelName) throws IOException {
+    return String.format("{\"op\": \"subscribe\", \"args\": [\"%s\"]}", channelName);
+  }
+
+  @Override
+  protected void handleMessage(JsonNode message) {
+    if (message.get("event") != null && "pong".equals(message.get("event").asText())) {
+      // ignore pong message
+      return;
+    }
+    super.handleMessage(message);
+  }
+
+  @Override
+  protected WebSocketClientHandler getWebSocketClientHandler(
+      WebSocketClientHandshaker handshaker,
+      WebSocketClientHandler.WebSocketMessageHandler handler) {
+    return new OkCoinNettyWebSocketClientHandler(handshaker, handler);
+  }
+
+  protected class OkCoinNettyWebSocketClientHandler extends NettyWebSocketClientHandler {
+
+    private final Logger LOG = LoggerFactory.getLogger(OkCoinNettyWebSocketClientHandler.class);
+
+    protected OkCoinNettyWebSocketClientHandler(
+        WebSocketClientHandshaker handshaker, WebSocketMessageHandler handler) {
+      super(handshaker, handler);
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) {
+      if (pingPongSubscription != null && !pingPongSubscription.isDisposed()) {
+        pingPongSubscription.dispose();
+      }
+      super.channelInactive(ctx);
+    }
+
+    @Override
+    public void channelRead0(ChannelHandlerContext ctx, Object msg) throws Exception {
```

Comment:
Looks like you can share the one in `OkCoinStreamingService` (just move it to a new file) and add that extra check for `pong` at the end; it's probably safe for both versions.